### PR TITLE
[Shop][Account] Marking an address as default

### PR DESCRIPTION
**Pull request by NeverResponse**
| Q               | A
| --------------- | ---
| Bug fix?        | no
| New feature?    | yes
| BC breaks?      | no
| Related tickets | -
| License         | MIT

### CODE REVIEW

File: src/Sylius/Bundle/CoreBundle/Form/Type/AddressChoiceType.php [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+    public function configureOptions(OptionsResolver $resolver)
+    {
+        $choices = function (Options $options) {
+            if (null === $options['customer']) {
+                $choices = $this->addressRepository->findAll();
+            } else {
+                $choices = $this->addressRepository->findBy(['customer' => $options['customer']]);
+            }
+
+            return $choices;
```
Comment (Zales0123) [inline src/Sylius/Bundle/CoreBundle/Form/Type/AddressChoiceType.php]: Why not:
```php
if (null === $options['customer']) {
    return $this->addressRepository->findAll();
}

return $this->addressRepository->findBy(['customer' => $options['customer']]);
```

File: src/Sylius/Bundle/UiBundle/Resources/translations/flashes.en.yml [hunk 2]
```diff
@@ -2,3 +2,4 @@ sylius:
     ui:
         the_channel_cannot_be_deleted_at_least_one_enabled_channel_is_required: 'The channel cannot be deleted. At least one enabled channel is required.'
         address_has_been_successfully_added: 'Address has been successfully added.'
+        address_has_been_set_as_default: 'Address has been set as default.'
```
Comment (Zales0123) [inline src/Sylius/Bundle/UiBundle/Resources/translations/flashes.en.yml]: Would you consider sorting this key alphabetically?

File: src/Sylius/Behat/Context/Ui/Shop/AddressBookContext.php [hunk 3]
```diff
@@ -105,6 +105,17 @@ public function iEditAddressOf($fullName)
     }
 
     /**
+     * @Given my default address is of :fullName
```
Comment (michalmarcinkowski) [inline src/Sylius/Behat/Context/Ui/Shop/AddressBookContext.php]: `Given` should be placed in `Setup/*Context` and should be done programmatically.

File: src/Sylius/Behat/Page/Shop/Account/AddressBook/IndexPage.php [hunk 4]
```diff
@@ -81,6 +81,31 @@ public function deleteAddress($fullName)
...
+    {
+        return !$this->hasElement('default_address');
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function getFullNameOfDefaultAddress()
+    {
+        return $this->getElement('default_address')->find('css', 'address > strong')->getText();
```
Comment (michalmarcinkowski) [inline src/Sylius/Behat/Page/Shop/Account/AddressBook/IndexPage.php]: `$this->getElement('default_address')->find()` can return `null` we should `Assert::notNull` before using `->getText()` method.

File: src/Sylius/Bundle/CoreBundle/Form/Type/Customer/CustomerDefaultAddressType.php [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+{
+    /**
+     * {@inheritdoc}
+     */
+    public function buildForm(FormBuilderInterface $builder, array $options)
+    {
+        $builder
+            ->add('defaultAddress', 'sylius_address_choice', [
+                'choices_as_values' => true,
+                'customer' => $options['customerId'],
```
Comment (michalmarcinkowski) [inline src/Sylius/Bundle/CoreBundle/Form/Type/Customer/CustomerDefaultAddressType.php]: We should pass `CustomerInterface` object to the form type.

File: src/Sylius/Bundle/CoreBundle/Form/Type/Customer/CustomerDefaultAddressType.php [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+        ;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function configureOptions(OptionsResolver $resolver)
+    {
+        $resolver->setRequired('customerId');
+        $resolver->setDefault('csrf_protection', false);
```
Comment (michalmarcinkowski) [inline src/Sylius/Bundle/CoreBundle/Form/Type/Customer/CustomerDefaultAddressType.php]: Do we need to do that?

**Comment by Zales0123** [comment]
I suppose it can be closed in favor of https://github.com/Sylius/Sylius/pull/6584

File: src/Sylius/Bundle/ShopBundle/Resources/views/Account/AddressBook/_defaultAddressForm.html.twig [hunk 7]
```diff
@@ -0,0 +1,8 @@
+{% set id = app.request.attributes.get('id') %}
+
+{{ form_start(form, {'action': path('sylius_shop_account_address_book_set_default', {'id': id}), 'method': 'PATCH'}) }}
+    <input hidden name="{{ form.defaultAddress.vars['full_name'] }}" value="{{ id }}" />
```
Comment (michalmarcinkowski) [inline src/Sylius/Bundle/ShopBundle/Resources/views/Account/AddressBook/_defaultAddressForm.html.twig]: We shouldn't add `name="{{ form.defaultAddress.vars['full_name'] }}"` just for Behat.

File: src/Sylius/Bundle/ShopBundle/Resources/config/routing/account/address_book.yml [hunk 8]
```diff
@@ -56,9 +56,31 @@ sylius_shop_account_address_book_delete:
...
+    methods: [GET, PATCH]
+    defaults:
+        _controller: sylius.controller.customer:updateAction
+        _sylius:
+            template: SyliusShopBundle:Account/AddressBook:_defaultAddressForm.html.twig
+            section: shop_account
+            form:
+                type: sylius_customer_default_address
+                options:
+                    customerId: 'expr:service("sylius.context.customer").getCustomer().getId()'
```
Comment (michalmarcinkowski) [inline src/Sylius/Bundle/ShopBundle/Resources/config/routing/account/address_book.yml]: No need for `.getId()`

**Comment by michalmarcinkowski** [comment]
Closing in favour of https://github.com/Sylius/Sylius/pull/6584

File: src/Sylius/Bundle/CoreBundle/Form/Type/Customer/CustomerDefaultAddressType.php (hunk 6 above)
Comment (NeverResponse) [inline src/Sylius/Bundle/CoreBundle/Form/Type/Customer/CustomerDefaultAddressType.php]: It's either this or rendering a hidden token input, but I'll go with the latter.

**Pull request closed without merging**